File: repertoire/concepts/data-integration/providers/structure_detector/json_schema_detector.ts
// JSON schema detector — infers schema from JSON objects or CSV content
// Detects field types, patterns (email, URL, date, UUID), cardinality, nullability

export const PROVIDER_ID = 'json_schema_detector';
export const PLUGIN_TYPE = 'structure_detector';

export interface DetectorConfig {
  options?: Record<string, unknown>;
  confidenceThreshold?: number;
}

export interface Detection {
  field: string;
  value: unknown;
  type: string;
  confidence: number;
  evidence: string;
}

interface FieldSchema {
  name: string;
  type: string;
  pattern?: string;
  nullable: boolean;
  count: number;
  totalSamples: number;
}

const PATTERN_MATCHERS: Array<{ name: string; regex: RegExp }> = [
  { name: 'email', regex: /^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}$/ },
  { name: 'url', regex: /^https?:\/\/\S+$/ },
  { name: 'uuid', regex: /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i },
  { name: 'iso_date', regex: /^\d{4}-\d{2}-\d{2}(T\d{2}:\d{2}:\d{2})?/ },
  { name: 'ipv4', regex: /^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$/ },
  { name: 'phone', regex: /^\+?\d[\d\s()-]{6,20}$/ },
];

function inferJsonType(value: unknown): string {
  if (value === null || value === undefined) return 'null';
  if (typeof value === 'string') return 'string';
  if (typeof value === 'number') return Number.isInteger(value) ? 'integer' : 'number';
  if (typeof value === 'boolean') return 'boolean';
  if (Array.isArray(value)) return 'array';
  if (typeof value === 'object') return 'object';
  return 'unknown';
}

function detectPattern(value: string): string | undefined {
  for (const matcher of PATTERN_MATCHERS) {
    if (matcher.regex.test(value)) return matcher.name;
  }
  return undefined;
}

function analyzeObjects(objects: Record<string, unknown>[]): FieldSchema[] {
  const fieldMap = new Map<string, {
    types: Map<string, number>;
    patterns: Map<string, number>;
    nullCount: number;
    count: number;
  }>();

  const totalSamples = objects.length;

  for (const obj of objects) {
    for (const [key, val] of Object.entries(obj)) {
      if (!fieldMap.has(key)) {
        fieldMap.set(key, { types: new Map(), patterns: new Map(), nullCount: 0, count: 0 });
      }
      const stats = fieldMap.get(key)!;
      stats.count++;

      const type = inferJsonType(val);
      stats.types.set(type, (stats.types.get(type) ?? 0) + 1);

      if (val === null || val === undefined) {
        stats.nullCount++;
      } else if (typeof val === 'string') {
        const pattern = detectPattern(val);
        if (pattern) {
          stats.patterns.set(pattern, (stats.patterns.get(pattern) ?? 0) + 1);
        }
      }
    }
  }

  const schemas: FieldSchema[] = [];
  for (const [name, stats] of fieldMap) {
    // Determine dominant type
    let dominantType = 'string';
    let maxCount = 0;
    for (const [type, count] of stats.types) {
      if (type !== 'null' && count > maxCount) {
        dominantType = type;
        maxCount = count;
      }
    }

    // Determine dominant pattern
    let pattern: string | undefined;
    for (const [pat, count] of stats.patterns) {
      if (count > (stats.count - stats.nullCount) * 0.7) {
        pattern = pat;
        break;
      }
    }

    schemas.push({
      name,
      type: dominantType,
      pattern,
      nullable: stats.nullCount > 0,
      count: stats.count,
      totalSamples,
    });
  }

  return schemas;
}

function parseCsvToObjects(text: string): Record<string, unknown>[] {
  const lines = text.split(/\r?\n/).filter(l => l.trim().length > 0);
  if (lines.length < 2) return [];

  const delimiter = lines[0].includes('\t') ? '\t' : ',';
  const headers = lines[0].split(delimiter).map(h => h.trim().replace(/^"|"$/g, ''));
  const objects: Record<string, unknown>[] = [];

  for (let i = 1; i < lines.length; i++) {
    const values = lines[i].split(delimiter).map(v => v.trim().replace(/^"|"$/g, ''));
    const obj: Record<string, unknown> = {};
    for (let j = 0; j < headers.length; j++) {
      const raw = values[j] ?? '';
      if (raw === '' || raw.toLowerCase() === 'null') {
        obj[headers[j]] = null;
      } else if (/^-?\d+$/.test(raw)) {
        obj[headers[j]] = parseInt(raw, 10);
      } else if (/^-?\d+\.\d+$/.test(raw)) {
        obj[headers[j]] = parseFloat(raw);
      } else if (/^(true|false)$/i.test(raw)) {
        obj[headers[j]] = raw.toLowerCase() === 'true';
      } else {
        obj[headers[j]] = raw;
      }
    }
    objects.push(obj);
  }

  return objects;
}

export class JsonSchemaDetectorProvider {
  detect(
    content: unknown,
    existingStructure: Record<string, unknown>,
    config: DetectorConfig
  ): Detection[] {
    const threshold = config.confidenceThreshold ?? 0.5;
    let objects: Record<string, unknown>[] = [];

    // Try parsing as JSON
    if (typeof content === 'string') {
      try {
        const parsed = JSON.parse(content);
        if (Array.isArray(parsed)) {
          objects = parsed.filter(item => typeof item === 'object' && item !== null);
        } else if (typeof parsed === 'object' && parsed !== null) {
          objects = [parsed];
        }
      } catch {
        // Try CSV parsing
        objects = parseCsvToObjects(content);
      }
    } else if (Array.isArray(content)) {
      objects = content.filter(item => typeof item === 'object' && item !== null);
    } else if (typeof content === 'object' && content !== null) {
      objects = [content as Record<string, unknown>];
    }

    if (objects.length === 0) return [];

    const schemas = analyzeObjects(objects);
    const detections: Detection[] = [];

    for (const schema of schemas) {
      const cardinality = schema.count / schema.totalSamples;
      let confidence = cardinality >= 1.0 ? 0.95 : 0.80;
      if (schema.pattern) confidence = Math.min(confidence + 0.03, 0.99);

      if (confidence < threshold) continue;

      detections.push({
        field: `schema.${schema.name}`,
        value: {
          type: schema.type,
          pattern: schema.pattern ?? null,
          nullable: schema.nullable,
          cardinality: Math.round(cardinality * 100) / 100,
          samples: schema.totalSamples,
        },
        type: 'schema_field',
        confidence,
        evidence: `Field "${schema.name}": ${schema.type}${schema.pattern ? ` (${schema.pattern})` : ''}, ${schema.nullable ? 'nullable' : 'required'}`,
      });
    }

    // Add overall schema detection
    if (detections.length > 0) {
      detections.push({
        field: 'schema',
        value: {
          fieldCount: schemas.length,
          sampleCount: objects.length,
          fields: schemas.map(s => s.name),
        },
        type: 'json_schema',
        confidence: 0.90,
        evidence: `Schema with ${schemas.length} fields from ${objects.length} sample(s)`,
      });
    }

    return detections;
  }

  appliesTo(contentType: string): boolean {
    return ['application/json', 'text/csv', 'text/tab-separated-values', 'application/x-ndjson'].includes(contentType);
  }
}

export default JsonSchemaDetectorProvider;

File: repertoire/concepts/data-integration/providers/structure_detector/json_schema_detector.rs
// JSON schema detector — infers schema from JSON objects or CSV content
// Detects field types, patterns (email, URL, date, UUID), cardinality, nullability

use regex::Regex;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::HashMap;

pub const PROVIDER_ID: &str = "json_schema_detector";
pub const PLUGIN_TYPE: &str = "structure_detector";

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DetectorConfig {
    pub options: Option<HashMap<String, Value>>,
    pub confidence_threshold: Option<f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Detection {
    pub field: String,
    pub value: Value,
    pub r#type: String,
    pub confidence: f64,
    pub evidence: String,
}

#[derive(Debug)]
pub enum DetectorError {
    ParseError(String),
    RegexError(String),
}

impl std::fmt::Display for DetectorError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            DetectorError::ParseError(msg) => write!(f, "Parse error: {}", msg),
            DetectorError::RegexError(msg) => write!(f, "Regex error: {}", msg),
        }
    }
}

fn infer_json_type(value: &Value) -> &'static str {
    match value {
        Value::Null => "null",
        Value::String(_) => "string",
        Value::Bool(_) => "boolean",
        Value::Number(n) => if n.is_f64() && n.as_i64().is_none() { "number" } else { "integer" },
        Value::Array(_) => "array",
        Value::Object(_) => "object",
    }
}

fn detect_pattern(value: &str) -> Option<&'static str> {
    let patterns: Vec<(&str, Regex)> = vec![
        ("email", Regex::new(r"^[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}$").unwrap()),
        ("url", Regex::new(r"^https?://\S+$").unwrap()),
        ("uuid", Regex::new(r"(?i)^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$").unwrap()),
        ("iso_date", Regex::new(r"^\d{4}-\d{2}-\d{2}(T\d{2}:\d{2}:\d{2})?").unwrap()),
        ("ipv4", Regex::new(r"^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$").unwrap()),
        ("phone", Regex::new(r"^\+?\d[\d\s()\-]{6,20}$").unwrap()),
    ];
    for (name, re) in &patterns {
        if re.is_match(value) { return Some(name); }
    }
    None
}

struct FieldStats {
    types: HashMap<String, usize>,
    patterns: HashMap<String, usize>,
    null_count: usize,
    count: usize,
}

fn analyze_objects(objects: &[&serde_json::Map<String, Value>]) -> Vec<(String, String, Option<String>, bool, usize, usize)> {
    let mut field_map: HashMap<String, FieldStats> = HashMap::new();
    let total = objects.len();

    for obj in objects {
        for (key, val) in *obj {
            let stats = field_map.entry(key.clone()).or_insert_with(|| FieldStats {
                types: HashMap::new(), patterns: HashMap::new(), null_count: 0, count: 0,
            });
            stats.count += 1;
            let t = infer_json_type(val);
            *stats.types.entry(t.to_string()).or_insert(0) += 1;

            if val.is_null() {
                stats.null_count += 1;
            } else if let Some(s) = val.as_str() {
                if let Some(pat) = detect_pattern(s) {
                    *stats.patterns.entry(pat.to_string()).or_insert(0) += 1;
                }
            }
        }
    }

    field_map.into_iter().map(|(name, stats)| {
        let dominant_type = stats.types.iter()
            .filter(|(t, _)| t.as_str() != "null")
            .max_by_key(|(_, c)| *c)
            .map(|(t, _)| t.clone())
            .unwrap_or_else(|| "string".to_string());

        let non_null = stats.count - stats.null_count;
        let pattern = stats.patterns.iter()
            .find(|(_, &c)| c as f64 > non_null as f64 * 0.7)
            .map(|(p, _)| p.clone());

        let nullable = stats.null_count > 0;
        (name, dominant_type, pattern, nullable, stats.count, total)
    }).collect()
}

fn parse_csv_to_json(text: &str) -> Vec<serde_json::Map<String, Value>> {
    let lines: Vec<&str> = text.lines().filter(|l| !l.trim().is_empty()).collect();
    if lines.len() < 2 { return vec![]; }

    let delimiter = if lines[0].contains('\t') { '\t' } else { ',' };
    let headers: Vec<String> = lines[0].split(delimiter)
        .map(|h| h.trim().trim_matches('"').to_string())
        .collect();

    let int_re = Regex::new(r"^-?\d+$").unwrap();
    let float_re = Regex::new(r"^-?\d+\.\d+$").unwrap();

    lines[1..].iter().map(|line| {
        let values: Vec<&str> = line.split(delimiter).collect();
        let mut obj = serde_json::Map::new();
        for (i, header) in headers.iter().enumerate() {
            let raw = values.get(i).unwrap_or(&"").trim().trim_matches('"');
            let val = if raw.is_empty() || raw.eq_ignore_ascii_case("null") {
                Value::Null
            } else if int_re.is_match(raw) {
                raw.parse::<i64>().map(Value::from).unwrap_or(Value::String(raw.to_string()))
            } else if float_re.is_match(raw) {
                raw.parse::<f64>().map(|f| serde_json::Number::from_f64(f).map(Value::Number).unwrap_or(Value::String(raw.to_string())))
                    .unwrap_or(Value::String(raw.to_string()))
            } else if raw.eq_ignore_ascii_case("true") {
                Value::Bool(true)
            } else if raw.eq_ignore_ascii_case("false") {
                Value::Bool(false)
            } else {
                Value::String(raw.to_string())
            };
            obj.insert(header.clone(), val);
        }
        obj
    }).collect()
}

pub struct JsonSchemaDetectorProvider;

impl JsonSchemaDetectorProvider {
    pub fn new() -> Self { Self }

    pub fn detect(
        &self,
        content: &str,
        _existing: &HashMap<String, Value>,
        config: &DetectorConfig,
    ) -> Result<Vec<Detection>, DetectorError> {
        let threshold = config.confidence_threshold.unwrap_or(0.5);

        // Try JSON parse first, then CSV
        let maps: Vec<serde_json::Map<String, Value>> = match serde_json::from_str::<Value>(content) {
            Ok(Value::Array(arr)) => arr.into_iter()
                .filter_map(|v| if let Value::Object(m) = v { Some(m) } else { None })
                .collect(),
            Ok(Value::Object(m)) => vec![m],
            _ => parse_csv_to_json(content),
        };

        if maps.is_empty() { return Ok(vec![]); }

        let refs: Vec<&serde_json::Map<String, Value>> = maps.iter().collect();
        let schemas = analyze_objects(&refs);
        let mut detections = Vec::new();
        let mut field_names = Vec::new();

        for (name, dtype, pattern, nullable, count, total) in &schemas {
            let cardinality = *count as f64 / *total as f64;
            let mut confidence = if cardinality >= 1.0 { 0.95 } else { 0.80 };
            if pattern.is_some() { confidence = (confidence + 0.03).min(0.99); }
            if confidence < threshold { continue; }

            field_names.push(name.clone());
            let mut val = serde_json::Map::new();
            val.insert("type".into(), Value::String(dtype.clone()));
            val.insert("pattern".into(), pattern.as_ref().map(|p| Value::String(p.clone())).unwrap_or(Value::Null));
            val.insert("nullable".into(), Value::Bool(*nullable));
            val.insert("cardinality".into(), serde_json::Number::from_f64((cardinality * 100.0).round() / 100.0).map(Value::Number).unwrap_or(Value::Null));
            val.insert("samples".into(), Value::Number((*total).into()));

            let pat_str = pattern.as_ref().map(|p| format!(" ({})", p)).unwrap_or_default();
            let null_str = if *nullable { "nullable" } else { "required" };

            detections.push(Detection {
                field: format!("schema.{}", name),
                value: Value::Object(val),
                r#type: "schema_field".into(),
                confidence,
                evidence: format!("Field \"{}\": {}{}, {}", name, dtype, pat_str, null_str),
            });
        }

        if !detections.is_empty() {
            let mut summary = serde_json::Map::new();
            summary.insert("fieldCount".into(), Value::Number(schemas.len().into()));
            summary.insert("sampleCount".into(), Value::Number(maps.len().into()));
            summary.insert("fields".into(), Value::Array(field_names.into_iter().map(Value::String).collect()));

            detections.push(Detection {
                field: "schema".into(),
                value: Value::Object(summary),
                r#type: "json_schema".into(),
                confidence: 0.90,
                evidence: format!("Schema with {} fields from {} sample(s)", schemas.len(), maps.len()),
            });
        }

        Ok(detections)
    }

    pub fn applies_to(&self, content_type: &str) -> bool {
        matches!(content_type, "application/json" | "text/csv" | "text/tab-separated-values" | "application/x-ndjson")
    }
}
